### PR TITLE
Combined PerceiverIO Encoder and Decoder

### PR DESCRIPTION
**Pull request by adisingh50**
Created `model.py` to combine the PerceiverIO encoder and decoder. Also created `test_model.py` to ensure that our PeceiverIO model can handle arbitrarily sized inputs. 

PerceiverIO Model:
- Input is of shape:  `[batch_size, num_pillars, feature_dimensionality]`
- Output is of shape: `[batch_size, num_pillars, 3]`

### CODE REVIEW

File: rasternet/perceiver_io/decoder.py [hunk 1]
```diff
@@ -41,12 +41,8 @@ def __init__(
         self.num_cross_attention_heads = num_cross_attention_heads
         self.dropout = dropout
 
-        self.output_query_dimensionality = self.output_query_shape[
-            1
-        ]  # number of columns in output query array.
-        self.latent_dimensionality = self.latent_shape[
-            1
-        ]  # number of columns in latent array.
+        self.output_query_dimensionality = self.output_query_shape[1]  # number of columns in output query array.
```
Comment (johnwlambert) [inline rasternet/perceiver_io/decoder.py:44]: nit: prefer `self.output_query_dim` for brevity

File: rasternet/perceiver_io/decoder.py [hunk 2]
```diff
@@ -41,12 +41,8 @@ def __init__(
...
+        self.latent_dimensionality = self.latent_shape[1]  # number of columns in latent array.
```
Comment (johnwlambert) [inline rasternet/perceiver_io/decoder.py:45]: same for `latent_dim`

File: rasternet/perceiver_io/model.py [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+        num_self_attention_layers_per_block: int,
+        dropout: float,
+        output_query_shape: Tuple[int, int],
+        final_output_dimensionality: int,
+    ) -> None:
+        """Initializes a PerceiverIO encoder and decoder.
+
+        Args:
+            input_dimensionality (int): dimensionality of input tensor.
+            latent_shape (Tuple[int, int]): shape of latent array, [num_latent_variables, latent_dimensionality].
```
Comment (johnwlambert) [inline rasternet/perceiver_io/model.py:31]: nit: no need to repeat type hint here (redundant w/ fn signature)

File: rasternet/perceiver_io/decoder.py (hunk 1 above)
Comment (adisingh50) [reply rasternet/perceiver_io/decoder.py:44]: yup

File: rasternet/perceiver_io/decoder.py (hunk 2 above)
Comment (adisingh50) [reply rasternet/perceiver_io/decoder.py:45]: yup

File: rasternet/perceiver_io/model.py (hunk 3 above)
Comment (adisingh50) [reply rasternet/perceiver_io/model.py:31]: yup

**Comment by johnwlambert** [comment]
yeah it includes it for training. inference could be a different story though

**Comment by anshulcgm** [comment]
> 

Don’t we want to design the module with training in mind though, i’ve usually seen softmax run after the forward pass is called. Just a tiny nitpick, it can easily be changed in the future but when we want to train we’re gonna have to remove the softmax call so that we don’t call it twice when we use cross entropy loss

**Comment by adisingh50** [comment]
Yeah, I forgot that Pytorch's Cross Entropy loss includes the softmax operation. I can remove that softmax line.
